### DISCUSSION TITLE
Vertical text detection and extraction?

### DISCUSSION THREAD

**Discussion by mllife** [mindee/doctr#1524] [Q&A] [answered]
I am working on extracting text, but the framework is unable to handle vertical text. For example - 
<img width="754" alt="Screenshot 2024-03-26 at 11 23 08 AM" src="https://github.com/mindee/doctr/assets/69006366/222fad0c-db96-4623-a0d1-921895fcc2b7">
Can someone help, if I am missing something?

**Comment by mllife** [comment]
@fg-mindee , please help.

**Reply by felixdittrich92** [reply, marked as answer]
Hi @mllife :wave:,

You need to pass `assume_straight_pages=False` to the `ocr_predictor` instance :)
I suggest to use the main branch instead of v0.8.1 and `fast_tiny` or `fast_base` as detection model (fast_base will be published today so at the moment of this answer only fast_tiny is pretrained available) - only with pytorch as backend available TF checkpoints will follow in the next weeks
![Screenshot from 2024-03-26 07-14-28](https://github.com/mindee/doctr/assets/23098070/c3042d37-3489-48b7-a66a-7c432f98cf14)

**Reply by felixdittrich92** [reply]
CC @odulcy-mindee the output is from `fast_base` if we can reproduce the quality with TF we could change the default det arch to fast which works in general more robust (also with rotations), needs less memory and is a bit faster compared to db_resnet50 (the current default)

**Comment by mllife** [comment]
Thanks @felixdittrich92 for your immediate response.I am able to get the "python-doctr-0.9.0a0" from the main branch, and "fast_base" works perfectly with the detection part, but the recognition is still bad on rotated text. I am using reco_arch = "parseq".
See from the image for reconstructed text. Are there any immediate plans to update recognition models as well for indentifying the rotated text? I am also very much impressed DocumentStructure class representation for the final result, but it is also lacking orientation attribute. Is that something that is doable in the upcoming release.
![reconstructed_page](https://github.com/mindee/doctr/assets/69006366/d8e6cbe5-8c13-4371-aac1-d43c08808284)

**Reply by felixdittrich92** [reply]
This looks not correct ^^ see:

```
model = ocr_predictor(
    det_arch="fast_base",
    reco_arch="parseq",
    pretrained=True,
    det_bs=8,
    reco_bs=1024,
    assume_straight_pages=False,
    straighten_pages=False,
    detect_orientation=True,
) #.cuda().half()  # uncomment if you have a GPU

result = model(doc)

result.show()
json_res = result.export()

print(result.render())

print(f"Rotation: {json_res["pages"][0]["orientation"]["value"]}")
```

render result:
```
Consolidated Financial Statements
Total 74,749 7,528 500 8,028 1,399 I (5,523) 14 I 78,667
72,892 10,174 (1.589) 8,585 3,211 (3,373) (5,233) 10 (1,343) 74,749
$
$
$
$
$
$
Non- in 31(b)) 1,524 118 209 327 I I (101) - (11) 1,739
2,090 258 (25) 233 - I (115) (684)(4) 1,524
$
$
$
$
controlling
interests subsidiaries (Note
$
$
equity holders 73,225 7,410 291 7,701 1,399 I (5,422) 14 11 76,928
9,916 (1.564) 8,352 3,211 (3,373) (5,118) 10 (659)
to
70,802
73,225
$
$
attributable
$
$
$
$
Total
Rotation: -90
```

**Reply by felixdittrich92** [reply]
But yeah we still work on the rotation topic especially on improvements for the crop orientation predictor and to handle upside down :)
And as you can see the results are mostly correct but the `.render()` fails complete

**Reply by mllife** [reply]
Will the orientation attribute be available with version 0.9 in the DocumentStructure output?

**Reply by felixdittrich92** [reply]
It is already you need only to add `detect_orientation=True` to the predictor :)
But i have faiced that the `fast` models are the most robust for rotations atm compared to `db` and `linknet` and the orientation angle detection depends on the detection model returned bin map

**Reply by mllife** [reply]
Yes,  but "detect_orientation=True" only gives page orientation. I am talking about the individual word bboxes and their orientation, which makes more sense.
For example:
**_'words':_** [{_'value': 'Total'_, 'confidence': 0.9999986886978149, 'geometry': ((0.18405330181121826, 0.0888671875), (0.18405330181121826, 0.0693359375), (0.19416359066963196, 0.0693359375), (0.19416359066963196, 0.0888671875)), **_'orientation': -90_**}}

**Reply by mllife** [reply]
Can you add information about the 4 tuples now in geometry, it was just 2 in DocumentStructure ((x_min, y_min), (x_max, y_max))?

**Reply by felixdittrich92** [reply]
Yeah that could be indeed a great addition :)

Yep with assume-straight_pages=False you get polygons ((x1, y1), (x2, y2), (x3, y3), (x4, y4)) back

```
for page in result.pages:
        h, w = page.dimensions

        for block in page.blocks:
            for line in block.lines:
                for word in line.words:
                    coords = word.geometry
                    if len(coords) == 4:
                        x_coords = [int(point[0] * w) for point in coords]
                        y_coords = [int(point[1] * h) for point in coords]

                        x1, x2, x3, x4, y1, y2, y3, y4 = x_coords + y_coords

**Comment by mllife** [comment]
Thanks, @felixdittrich92. I hope the text/word orientation can be added in the upcoming version as it will help a lot of people.

**Reply by felixT2K** [reply]
Hi @mllife short question about the "word orientation".
What we could include is the general orientation of the text box ([0, 90, 180, 270] from the crop orientation classifier) is this what you mean or do you mean the orientation of the word in the crop ?

**Reply by mllife** [reply]
yes, as mentioned in the image example [image](https://private-user-images.githubusercontent.com/23098070/316747356-c3042d37-3489-48b7-a66a-7c432f98cf14.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE5NTA1MDQsIm5iZiI6MTcxMTk1MDIwNCwicGF0aCI6Ii8yMzA5ODA3MC8zMTY3NDczNTYtYzMwNDJkMzctMzQ4OS00OGI3LWE2NmEtN2M0MzJmOThjZjE0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDAxVDA1NDMyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI0YzM1YmNjYzM0MjNhZTk5YzM2NzdjMjVmZTU4Mjk1ZmJhN2MzYjMzZjdiYTU4ZWJkN2JhZDgyZjk0YzkwOWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2xPySBWV0v6v9cLLBbAWrCshSg6Qbahxg6F3UGGlPRI)
the text total is rotated by 90 with respect to the page, if it can be added into the document structure with each word it will be really helpful.
@felixT2K , thanks for looking into it. waiting for the next update to include this feature.

**Reply by Muaz08** [reply]
@felixdittrich92 I hope you are doing well. I wonder is there any way we could get the response of the doctr in the csv, excel and html that retains the same table structure as it was in the image. Thanks.
